Feat/#75

jyj1289 opened the pull request:
## 🎫 관련 이슈
[//]: # (다음 키워드를 사용하면 해당 PR을 머지할 때 자동으로 이슈를 닫을 수 있습니다.)
[//]: # (keyword: close|closes|closed|resolve|resolves|resolved|fix|fixes|fixed)
[//]: # (예시: close #1)

close #75

<br>

## 📄 개요
[//]: # (작업 내용을 간단히 요약해서 적습니다.)
[//]: # (예시: 유저 회원가입 기능을 만들었습니다.)
>그룹별 메시지 보내기 기능을 만들었습니다. 

<br>

## 🔨 작업 내용
[//]: # (작업 내용을 자세하게 적습니다.)
[//]: # (붙임표 "-" 을 사용해서 목록을 만듭니다.)
[//]: # (예시: 유저 회원가입 API를 만들었습니다.)


- 메시지 보내기 API를 만들었습니다. 
- SendMessageService에 execute 메서드를 오버로딩했습니다.


<br>

## 🏁 확인 사항
[//]: # (PR을 보내기 전 다음 사항을 확인해주세요.)
[//]: # (해당 사항을 모두 이행해야 머지할 수 있습니다.)
[//]: # (- [x] 를 사용해서 완료로 표시할 수 있습니다.)

- [x] 테스트를 완료했나요?
- [x] API 문서를 작성했나요?
- [ ] 코드 컨벤션을 준수했나요?
- [x] 불필요한 로그, 주석, import 등을 삭제했나요?

<br>

## 🙋🏻 덧붙일 말
[//]: # (다음 사항이 있다면 적어주세요.)
[//]: # (PR에 대한 추가 설명)
[//]: # (중점적으로 리뷰받고 싶은 부분)
[//]: # (기타 등등)

파일 구조가 정확한지, 역할에 맞게 코드를 작성했는지를 중점적으로 리뷰받고 싶습니다!

jyj1289 committed:
feat : MessageController
- 메시지를 그룹별로 보낼 수 있는 메세지 컨트롤러를 만들었습니다.

jyj1289 committed:
feat(#75):send message by group service
- 메시지를 그룹별로 보낼 수 있는 메세시를 보낼 수 있는 서비스를 만들었습니다.

jyj1289 committed:
docs:add cors url
- cors에 마루 버셀 url을 추가했습니다.

jyj1289 committed:
feat(#75):MessageControllerTest
- MessageControllerTest를 만들었어요.

jyj1289 committed:
docs(#75):MessageController asciidocs
- MessageController에 대한 아스키 독스를 만들었어요

jyj1289 committed:
fix(#75) : remove verify()
- 횟수 검증을 할 필요가 없어서 verify() 호출 코드를 제거했어요

jyj1289 committed:
refactor(#75) : 주석 제거
- 불필요한 주석을 제거했어요

jyj1289 committed:
refactor(#75) : 중복 코드 제거
- 불필요한 중복 코드를 제거했어요

gimhanul commented on src/main/java/com/bamdoliro/maru/infrastructure/message/SendMessageService.java:
```diff
@@ -28,6 +35,24 @@ public void execute(String to, String text, String title) {
+            phoneNumberList.add(form.getUser().getPhoneNumber());
+        }
+
+        for(String phoneNumber : phoneNumberList){
```
P4) 습관적 임포트 정리 + 습관적 자동 정렬 해주세요!!

gimhanul commented on src/main/java/com/bamdoliro/maru/presentation/message/MessageController.java:
```diff
@@ -0,0 +1,28 @@
+    public void sendMessage(
+            @AuthenticationPrincipal(authority = Authority.ADMIN) User user,
+            @RequestBody @Valid SendMessageRequest request
+            ){
```
P4) 여기두 습관적 자동 정렬

gimhanul commented on src/main/java/com/bamdoliro/maru/infrastructure/message/SendMessageService.java:
```diff
@@ -28,6 +35,24 @@ public void execute(String to, String text, String title) {
 
+    public void execute(SendMessageRequest request){
+        List<Form> formList = formRepository.findByStatus(request.getStatus());
+        List<String> phoneNumberList = new ArrayList<>();
+        List<Message> messageList = new ArrayList<>();
+
+        for (Form form : formList){
+            phoneNumberList.add(form.getUser().getPhoneNumber());
+        }
+
+        for(String phoneNumber : phoneNumberList){
+            Message message = createMessage(phoneNumber, request.getText());
+            message.setSubject(request.getTitle());
+            messageList.add(message);
+        }
+
+        sendManyMessage(messageList);
```
P2) stream 사용하면 코드가 훨씬 간결해질 거 같아요!

gimhanul commented on src/main/java/com/bamdoliro/maru/infrastructure/message/SendMessageService.java:
```diff
@@ -28,6 +35,24 @@ public void execute(String to, String text, String title) {
         sendOneMessage(message);
     }
 
+    public void execute(SendMessageRequest request){
```
P1) 이 메서드가 올 위치는 여기가 아닌 거 같아요
infrastructure라는 패키지는 말그대로 외부 시스템과 연결된 곳인데, 이 메서드 내부에서 하고 있는 일은 usecase와 다름 없습니다

usecase내부에서 phoneNumbers와 메시지를 세팅하고 해당 서비스를 호출하는 방식으로 바꿔주세요!

gimhanul commented on src/main/java/com/bamdoliro/maru/infrastructure/message/SendMessageService.java:
```diff
@@ -46,4 +71,13 @@ private void sendOneMessage(Message message) {
         }
     }
+
+    private void sendManyMessage(List<Message> messageList){
```
P5)

```suggestion
    private void sendManyMessages(List<Message> messageList){
```

gimhanul commented on src/main/resources/banner.txt:
```diff
@@ -6,7 +6,7 @@
 "`-0-0-'"`-0-0-'"`-0-0-'"`-0-0-'"`-0-0-'"`-0-0-'"`-0-0-'"`-0-0-'"`-0-0-'
 
-Contributors - @gimhanul, @hsem4717
+Contributors - @gimhanul, @hsem4717, @jyj1289, @cabbage16
```
👏 👏 👏 👏

gimhanul commented on src/test/java/com/bamdoliro/maru/presentation/message/MessageControllerTest.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @Test
+    void 원서가_접수된_학생들에게_메시지를_보낸다() throws Exception{
```
P1) 기획적 니즈가 잘 반영된 것인지 담당 선생님께 찾아가서 여쭤보면 좋을 거 같아요
원서 상태를 동적으로 해서 메시지를 보내고 싶은 것일 수도 있고, 현재 개발된 대로 원서가 접수된 학생들에 한해서만 보낼 수도 있는 거니까요

생각보다 기획적 내용 확인은 개발자들에게 중요하니 꼭 확인해주세요
다 갈아엎고 다시 개발해야할지도....

gimhanul reviewed (changes requested):
수고하셨습니다!

jyj1289 committed:
refactor(#75) : 자동 정렬
- 코드를 정렬했어요

jyj1289 committed:
feat(#75) : SendMessageUseCase
- SendMessageUseCase를 만들었어요

jyj1289 committed:
refact(#75) : MessageUseCase 사용
- execute에서 phoneNumberList를 생성하던 것을 UseCase로 이전했어요.
- 기존 for문을 stream으로 변환했어요.

jyj1289 committed:
refact(#75) : MessageUseCase 사용
- SendMessageService를 사용하던 코드들을 SendMessageUseCase를 사용하도록 변환했어요.

gimhanul commented on src/main/java/com/bamdoliro/maru/application/message/SendMessageUseCase.java:
```diff
@@ -0,0 +1,24 @@
+
+    public void execute(SendMessageRequest request) {
+        List<Form> formList = formRepository.findByStatus(request.getStatus());
+        List<String> phoneNumberList = formList.stream().map(form -> form.getUser().getPhoneNumber()).toList();
```
P4) stream 컨벤션입니다!

```suggestion
        List<String> phoneNumberList = formList.stream()
                                    .map(form -> form.getUser().getPhoneNumber())
                                    .toList();
```

gimhanul commented on src/main/java/com/bamdoliro/maru/infrastructure/message/SendMessageService.java:
```diff
@@ -38,9 +49,16 @@ private Message createMessage(String to, String text) {
+
+    private void sendManyMessages(List<Message> messageList) {
+        try {
+            messageService.send(messageList, false, false);
```
요건 그냥 질문인데, false 두 개는 무슨 의미일까용?

jyj1289 replied:
첫 번째 false는 번호 중복을 허용하는지, 두 번째 false는 발송한 메시지 list를 볼 것인지 확인하는 기능입니다!

gimhanul commented on src/main/java/com/bamdoliro/maru/infrastructure/message/SendMessageService.java:
```diff
@@ -28,6 +32,13 @@ public void execute(String to, String text, String title) {
+    public void execute(List<String> phoneNumberList, String text, String title) {
+        List<Message> messageList = phoneNumberList.stream()
+                .map(phoneNumber -> createMessage(phoneNumber, text))
+                .peek(message -> message.setSubject(title)).toList();
```
P2) peek라는 메서드가 중간 연산으로 사용될 때만 동작하는 메서드라고 하네요
여러모로 기대한대로 동작하지 않는 기능이라 많은 사람들이 사용을 기피하는 친구래요
잘 동작하는지 디버깅 한 번 해보시면 좋을 거 같아요

안 한다면 forEach로 바꿔주시구욥

jyj1289 replied:
현재는 잘 작동하고 있지만 혹시 forEach로 바꾸는 편이 더 나을까요?

gimhanul replied:
잘 동작하면 괜찮을 거 같아용~

gimhanul commented on src/test/java/com/bamdoliro/maru/presentation/message/MessageControllerTest.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    @Test
+    void 원서가_접수된_학생들에게_메시지를_보낸다() throws Exception {
```
P2) status에 따라 보내는 걸로 변경되었으니, 테스트케이스별로 이름도 다 변경해주시면 좋을 거 같아요
유지보수성이 올라갑니다

jyj1289 committed:
refactor(#75) : stream 컨벤션 사용
- stream 컨벤션을 사용했어요.

jyj1289 committed:
refactor(#75) : sendMessage 메서드 이름 변경 및 엔드포인트 변경
- 원서 상태별로 메시지 보내기와 원서 타입별로 메시지 보내기 기능을 분리하기 위해 이름과 엔드포인트를 변경했어요.

jyj1289 committed:
test(#75) : 엔드포인트 변경 및 BDDMockito 메서드로 변경
- BDD관점으로 보기 위해서 BDDMockito메서드로 변경했어요.
- UseCaseTest에서 status별로 테스트하기 위해서 컨트롤러 테스트를 수정했어요.

jyj1289 committed:
test(#75) : status별 단위 테스트 작성
- SendMessageUseCaseTest를 만들었어요.

gimhanul reviewed (approved):
마지막으로 한 번 더 확인하고 머지해주세용~ 수고하셨습니다!
